### PR TITLE
feat(AI): Ships that run out of ammo will try to flee

### PR DESCRIPTION
**Pull request by TomGoodIdea**
**Bugfix:** This PR addresses issue #7585

## Fix Details
Ships that have only ammo-using weapons installed and have run out of ammo will try to flee the system instead of circling their target pointlessly.

## Testing Done
Went to a pirate system and checked.

### CODE REVIEW

**Comment by RisingLeaf** [comment]
Actually it's intended behaviour for these ships to be suicidal as it otherwise makes certain missions really hard to finish.

Furthermore it would be disappointing if you have a good fight and then the enemy runs away

**Comment by bene-dictator** [comment]
> Actually it's intended behaviour for these ships to be suicidal as it otherwise makes certain missions really hard to finish.
> 
> Furthermore it would be disappointing if you have a good fight and then the enemy runs away

It doesn't make sense for ships to be suicidal if they have no weapons, it just makes capturing easier. Pirates should try to survive, not be destroyed, even if they were originally implemented to let the player get ships for free. Mission ships should be accounted for by a different way I think.

**Comment by RisingLeaf** [comment]
> Ships with staying or daring personality are unaffected. Regarding enemies running away, it's the same situation as when the ship is `badly damaged`.

"badly damaged" ships should be way easier to hunt down

**Comment by bene-dictator** [comment]
I mean, watch the video I give in #7585. It's not really a fight, I'm just kiting the ship until it runs out of ammunition. That's not a good fight, even if it's good tactics. That ship doesn't have anything to fight back with, so it should run away. But if a ship has no ammo and has other weapons they can shoot with, they should use those weapons and not run away.

For stories like the FW, I understand that the aim is to disable or destroy ships and letting them flee would break them. But that's why Rainmakers have energy blasters. Even if they don't, the fact that they're not in the system they're meant to be occupying should be good enough, and could even open new paths for story writing. 

I'd like to see some testing with this PR in a playthrough of the Free Worlds storyline too.

**Comment by quyykk** [comment]
> Actually it's intended behaviour for these ships to be suicidal as it otherwise makes certain missions really hard to finish.

Do you have an example? Presumably this problems already exists for ships with low health no?

**Comment by TomGoodIdea** [comment]
It seems that Rainmaker Mark II (currently using only Sidewinders) is capable of having one energy blaster.

File: source/AI.cpp [hunk 1]
```diff
@@ -651,12 +651,33 @@ void AI::Step(const PlayerInfo &player, Command &activeCommands)
...
+					neeedsAmmo = true;
+			}
+		}
+
+		// Run away if your hostile target is not disabled
+		// and you are either badly damaged or have run out of ammo.
 		// Player ships never stop targeting hostiles, while hostile mission NPCs will
 		// do so only if they are allowed to leave.
 		const bool shouldFlee = (personality.IsFleeing() ||
 			(!personality.IsDaring() && !personality.IsStaying()
```
Comment (quyykk) [inline source/AI.cpp:679]: The daring personality check should only apply to the low health check. If a daring ship needs ammo it should IMO still flee.

**Review by quyykk — APPROVED** [review]

**Review by EjoThims — APPROVED** [review]
I don't really understand the code side of things super well, even in reading, but since I was asked, I'll weigh in.

If, as I understand it, this is about fleeing the system for ships that can leave the system and not changing the AI missile boat behavior of closing to attack for ships that can't (like mission targets and tribute fleet ships), it is a good change indeed.

File: source/AI.cpp [hunk 2]
```diff
@@ -651,12 +651,33 @@ void AI::Step(const PlayerInfo &player, Command &activeCommands)
...
+			{
+				const Outfit *ammo = weapon->Ammo();
+				if(!ammo || it->OutfitCount(ammo))
+				{
+					// This ship has at least one usable weapon.
+					needsAmmo = false;
+					break;
+				}
+				else
+					needsAmmo = true;
```
Comment (warp-core) [inline source/AI.cpp:670]:
```suggestion
		bool needsAmmo = true;
		for(const Hardpoint &hardpoint : it->Weapons())
		{
			const Weapon *weapon = hardpoint.GetOutfit();
			if(weapon && !hardpoint.IsAntiMissile())
			{
				const Outfit *ammo = weapon->Ammo();
				if(!ammo || it->OutfitCount(ammo))
				{
					// This ship has at least one usable weapon.
					needsAmmo = false;
					break;
				}
```

File: source/AI.cpp [hunk 3]
```diff
@@ -651,12 +651,33 @@ void AI::Step(const PlayerInfo &player, Command &activeCommands)
...
+		// Run away if your hostile target is not disabled
+		// and you are either badly damaged or have run out of ammo.
 		// Player ships never stop targeting hostiles, while hostile mission NPCs will
 		// do so only if they are allowed to leave.
-		const bool shouldFlee = (personality.IsFleeing() ||
-			(!personality.IsDaring() && !personality.IsStaying()
-			&& healthRemaining < RETREAT_HEALTH + .25 * personality.IsCoward()));
+		const bool shouldFlee = (personality.IsFleeing()
+				|| (!personality.IsStaying() && ((!personality.IsDaring()
+				&& healthRemaining < RETREAT_HEALTH + .25 * personality.IsCoward()) || needsAmmo)));
```
Comment (warp-core) [inline source/AI.cpp:680]: Refactor:
```suggestion
		const bool lowHealth = healthRemaining < RETREAT_HEALTH + .25 * personality.IsCoward();
		const bool shouldFlee = personality.IsFleeing()
				|| (!personality.IsStaying()
				&& ((!personality.IsDaring() && lowHealth) || needsAmmo));
```

File: source/AI.cpp [hunk 4]
```diff
@@ -651,12 +651,33 @@ void AI::Step(const PlayerInfo &player, Command &activeCommands)
...
+				if(!ammo || it->OutfitCount(ammo))
+				{
+					// This ship has at least one usable weapon.
+					needsAmmo = false;
+					break;
+				}
+				else
+					needsAmmo = true;
+			}
+		}
```
Comment (warp-core) [inline source/AI.cpp:672]: You should determine this in `ShipAICache::Calibrate()` and provide a method in that class to retrieve the result.

Comment (TomGoodIdea) [reply source/AI.cpp]: Is it OK now?

File: source/AI.cpp [hunk 5]
```diff
@@ -651,12 +651,15 @@ void AI::Step(const PlayerInfo &player, Command &activeCommands)
...
 		// Player ships never stop targeting hostiles, while hostile mission NPCs will
 		// do so only if they are allowed to leave.
-		const bool shouldFlee = (personality.IsFleeing() ||
-			(!personality.IsDaring() && !personality.IsStaying()
-			&& healthRemaining < RETREAT_HEALTH + .25 * personality.IsCoward()));
+		const bool lowHealth = healthRemaining < RETREAT_HEALTH + .25 * personality.IsCoward();
+		const bool needsAmmo = it->GetAICache().NeedsAmmo();
+		const bool shouldFlee = personality.IsFleeing()
+				|| (!personality.IsStaying()
+				&& ((!personality.IsDaring() && lowHealth) || needsAmmo));
```
Comment (quyykk) [inline source/AI.cpp:662]: Can you move this in a separate anonymous namespace function? and fix the conflicts and then this should be good to go :)

Comment (TomGoodIdea) [reply source/AI.cpp:662]: Do you mean `shouldFlee` or the fleeing procedure?

Comment (quyykk) [reply source/AI.cpp:730]: `shouldFlee`

Comment (warp-core) [reply source/AI.cpp:730]: Do what StayOrLinger() does for shouldStay.

**Review by quyykk — APPROVED** [review]

**Pull request merged**